### ISSUE TITLE
[Feature] Define DeepSeek V4 MTP model contract and model tests

### ISSUE THREAD

**Issue by high-cloud** [hw-native-sys/pypto-lib#694]
### Summary

Define the DeepSeek V4 MTP model-level contract in `pypto-lib` and add model tests for the target verification path and MTP draft model.

This issue intentionally scopes `pypto-lib` to model implementation responsibilities only. Full speculative decoding behavior, including scheduling, rejection sampling, request-state update, and end-to-end serving validation, should be owned and validated by the inference framework.

Related: #678

### Motivation / Use Case

DeepSeek V4 MTP requires a clear model contract between the target model, MTP draft model, and inference framework.

For `pypto-lib`, the important part is not to implement the whole speculative decoding engine, but to expose the model primitives that a framework needs:

1. target decode/prefill outputs required by MTP;
2. target verification forward with `Decode_Seq > 1`;
3. a complete MTP draft model path;
4. model tests that validate the input/output boundaries and cache overwrite semantics needed after draft rejection.

Without this contract, it is easy to validate only isolated fragments such as `mtp_projection`, while still missing the actual model behavior required by MTP inference.

#### Scope

In scope for `pypto-lib`:

- target `decode_fwd` / `prefill_fwd` contract required by MTP;
- target verification forward with `Decode_Seq > 1`;
- MTP draft model implementation;
- model-level tests for MTP inputs/outputs, hidden-state contract, and cache overwrite behavior at the model-kernel level.

Out of scope for `pypto-lib`:

- full speculative decoding scheduler;
- rejection sampling policy;
- serving request-state update;
- end-to-end acceptance-rate / throughput validation;
- framework-level correctness comparison with and without MTP.

Those should be validated by the inference framework.

#### Expected MTP flow

```mermaid
flowchart TD
    A["Target decode or prefill"] --> B["Sampled token"]
    A --> C["Target pre-hc-head residual hidden"]

    B --> D["MTP draft model"]
    C --> D

    D --> E["Draft tokens"]
    D --> F["MTP residual hidden for next draft step"]

    B --> G["Target verification forward"]
    E --> G

    G --> H["Target logits for sampled and draft tokens"]
    H --> I["Inference framework rejection sampling"]
    I --> J["Framework updates effective seq_len and request state"]
```

`pypto-lib` is responsible for `A`, `C`, `D`, and `G`.
The inference framework is responsible for `I` and `J`.

#### 1. Target forward should support MTP verification sequence length

Current DeepSeek V4 decode configuration appears to use:

```python
DECODE_SEQ = 1
```

For MTP verification, the target model needs to verify:

```text
last sampled token + draft tokens
```

Example with one draft token:

```text
committed context: A B C D E
last sampled token: F
MTP draft token: G'

target verification input: F G'
Decode_Seq = 2
```

Therefore `decode_fwd` needs to support at least:

```text
Decode_Seq = 1 + num_draft_tokens
```

For model tests, we should at minimum cover:

```text
Decode_Seq = 1
Decode_Seq = 2
```

and preferably make the implementation shape-parametric enough to cover larger `Decode_Seq` if the framework enables more speculative tokens.

#### 2. Target forward should expose pre-`hc_head` hidden if MTP needs it

DeepSeek V4 MTP draft consumes the target model's pre-`hc_head` residual hidden, not the final post-`hc_head` hidden.

Current `decode_fwd` / `prefill_fwd` appear to run:

```text
decoder layers -> hc_head -> final_norm -> lm_head/logits
```

We need to confirm whether the current kernels expose the intermediate tensor before `hc_head`.

If not, we need one of the following contracts:

Option A:

```text
decode_fwd / prefill_fwd optionally output pre_hc_hidden
```

Option B:

```text
provide a separate target-hidden forward path that returns pre_hc_hidden
```

Required shape should be equivalent to:

```text
[num_tokens, hc_mult * hidden_size]
```

For decode, this is needed after producing the last sampled token. For prefill, this may also be needed if the inference framework wants to start MTP immediately after prefill.

Open question:

```text
Is pre-hc-head hidden strictly required by the intended serving framework,
or can the framework obtain the correct MTP input through another model path?
```

Based on vLLM's DeepSeek V4 MTP design, this hidden contract is required.

#### 3. Provide a complete MTP draft model

Current code appears to have MTP-related fragments, such as `mtp_projection`, and some attention/cache paths with `mtp_kv_overlay`, but not a complete integrated MTP draft model.

The expected MTP draft model should include:

```text
input_ids / positions
target previous_hidden_states
token embedding
enorm / hnorm
e_proj / h_proj
MTP decoder layer
MTP attention / MoE / hc_post path as required
return next pre-hc residual hidden
compute draft logits via hc_head + shared head
```

The draft model forward should support chaining:

```text
step 0:
  previous_hidden = target pre_hc_hidden(F)
  input token = F
  output hidden = pre_hc_hidden(G')
  output logits -> sample G'

step 1:
  previous_hidden = pre_hc_hidden(G')
  input token = G'
  output hidden = pre_hc_hidden(H')
  output logits -> sample H'
```

If DeepSeek V4 has multiple MTP layers, the implementation should define how `spec_step_idx` selects the layer, for example:

```text
mtp_layer = spec_step_idx % num_mtp_layers
```

#### 4. Model test requirements

Add model tests for the following cases.

##### Decode sequence length

Compile/run target decode with:

```text
Decode_Seq = 1
Decode_Seq = 2
```

The `Decode_Seq = 2` case models target verification of:

```text
last sampled token + one draft token
```

This does not need serving integration. A model test can construct the two-token verification input directly.

##### Target pre-`hc_head` hidden

If the contract requires pre-`hc_head` hidden output, add a model test that checks:

```text
decode_fwd input -> pre_hc_hidden output
```

against a torch/golden reference at the exact boundary:

```text
after final decoder layer / hc_post
before hc_head
```

The test should ensure the output is not accidentally post-`hc_head` or post-final-norm hidden.

##### MTP draft single-step

Add a test for one MTP draft step.

Inputs:

```text
input_ids
positions
previous_hidden_states = target pre_hc_hidden
MTP weights
```

Outputs:

```text
next pre_hc_hidden
draft logits
```

Validate both hidden and logits against a torch/golden reference.

##### MTP draft multi-step chaining

Add a test where the output hidden from step 0 is used as input hidden for step 1:

```text
target pre_hc_hidden(F) -> MTP step 0 -> pre_hc_hidden(G')
pre_hc_hidden(G')       -> MTP step 1 -> pre_hc_hidden(H')
```

This validates that the draft model returns the correct hidden boundary for chaining.

##### Target verification model test

Add a target decode test with `Decode_Seq = 2`:

```text
input tokens: F G'
positions:    p p+1
```

The model test should verify:

```text
logits after F
logits after G'
KV writes for both positions
```

This is a model-level verification test only. It does not need to implement rejection sampling.

##### KV/cache overwrite contract

Even though rejection handling is framework-owned, model kernels must support overwrite semantics.

Construct a model-level test simulating:

```text
Step 1 target verification:
  input: F G'
  writes KV/cache at positions p and p+1

Framework rejects G':
  effective sequence length rolls back to p+1

Step 2 target decode:
  input: G
  position: p+1
```

Expected model-level property:

```text
KV/cache state for position p+1 is overwritten by G
```

If compressed KV or compressor state is involved, the same overwrite property must hold for:

```text
normal KV
compressed KV
compressor intermediate state
indexer/compressor metadata-derived slots
```

This test does not need to test framework rejection logic. It only needs to prove that when the framework reuses the same logical position/slot, the model implementation overwrites stale data correctly.

##### Compression-boundary overwrite

Add a targeted case where rejected `G'` lands on a compression boundary.

Example:

```text
G' position triggers compressed KV write
G' is logically rejected by the framework
G is later written at the same position
```

Expected:

```text
compressed slot for G overwrites compressed slot previously written by G'
```

This is important because append-only compression state would be unsafe for MTP rejection. The model implementation should be compatible with logical-position-based overwrite.

### Alternatives Considered

One alternative is to put the full speculative decoding loop into `pypto-lib`. That would make the model implementation harder to validate independently and would duplicate responsibilities that should belong to the inference framework.

Another alternative is to only validate isolated MTP fragments such as `mtp_projection`. That is too narrow: it does not prove that the target hidden boundary, draft chaining, logits path, and target verification shape are compatible with real MTP inference.

The proposed approach keeps `pypto-lib` focused on model primitives and model tests, while leaving end-to-end speculative decode validation to the inference framework.

### Additional Context

Current gaps to resolve:

| Area | Expected | Current Gap |
| --- | --- | --- |
| Decode sequence length | Support `Decode_Seq > 1` for target verification | Current DeepSeek V4 config appears fixed at `DECODE_SEQ = 1` |
| Target hidden contract | Expose pre-`hc_head` hidden if MTP requires it | Need confirm `decode_fwd` / `prefill_fwd` can output this boundary |
| MTP draft model | Integrated draft model forward + logits | Current code appears fragmented around `mtp_projection` and related pieces |
| MTP chaining | Return pre-`hc_head` residual for next draft step | Need explicit implementation and test |
| Verification forward | Target decode can run `last sampled + draft tokens` | Requires `Decode_Seq = 2+` test |
| KV overwrite | Rejected-token cache writes can be overwritten by real token at same position | Need model-level overwrite test |
| Compressed KV overwrite | Compression-boundary rejected token does not poison later decode | Need dedicated compression-boundary test |

Acceptance criteria:

1. Document the DeepSeek V4 MTP model-level contract in `pypto-lib`.
2. Support target decode verification with `Decode_Seq > 1`, at least `Decode_Seq = 2`.
3. Decide and document whether `decode_fwd` / `prefill_fwd` must output pre-`hc_head` hidden.
4. If required, implement pre-`hc_head` hidden output or an equivalent target-hidden path.
5. Implement a complete MTP draft model path, not only isolated projection fragments.
6. Add model tests for:
   - `Decode_Seq = 2` target verification;
   - pre-`hc_head` hidden output;
   - single-step MTP draft;
   - multi-step MTP chaining;
   - KV overwrite after simulated rejection;
   - compressed-KV overwrite on compression boundary.
7. Leave end-to-end speculative decode validation to the inference framework.
